### PR TITLE
fix: target blank removed from anchor tag

### LINKED ISSUE #4716: Allow "target=_blank" in the note of classDiagram
### Description

There is no attr 'target' in the rendered page. When using classDiagram, links are required to open in new tabs.
```
classDiagram
    class test {
    }
    note for test "<a href='https://mermaid.js.org/' target='_blank'><code>mermaid</code></a>"
```

```mermaid
classDiagram
    class test {
    }
    note for test "<a href='https://mermaid.js.org/' target='_blank'><code>mermaid</code></a>"
```

### Steps to reproduce

Using the live editor: https://mermaid.live/edit#pako:eNpNkMtuhDAMRX8FecMGAYHySMSwapdddVdFqtwhPKZAUDBSp4h_bwZQVW9sH1v3Wl7hqisFAq49zvNzh43BQY6OjZ04pGZy1oNsRxo1KafW5phJKNBpjaovbks0zSIIBmUG7Cr_NvvaNIHrEJpG0cX9-Oxx_HLLc6EIHuZlEWApATw4sb1mN5RArRqUBGHLStW49CRBjptdxYX02328giCzKA-WqUJS5_0gauxnSyccQazwDSJmuR_mYRixjGcszVIP7iAsjHjIo4QlcR4lYRxtHvxobRVCP-VPCY_yOE5YznPGd7n3ffjf86XqSJs_S7W3r-dTH2n7BaVdb7M

### Screenshots

_No response_

### Code Sample

_No response_

### Setup

- Mermaid version: 10.3.0
- Browser and Version: [Chrome 115.0.5790.171]

### Suggested Solutions

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by REVERB283**
## :bookmark_tabs: Summary

target=_blank was getting sanitized

Resolves #4716 

## :straight_ruler: Design Decisions

I've added a Dompurify hook that will add target="_blank" and rel="noopener noreferrer" back to the elements which will have target property in them. While debugging I found that Dompurify when sanitizing the input string removes target property for security reasons but if we really need that I found this could be an option.

### CODE REVIEW

File: packages/mermaid/src/diagrams/common/common.ts
Comment (REVERB283) [reply packages/mermaid/src/diagrams/common/common.ts:35]: Okay thanks for the pointers. I'll make those changes and also these checks are failing because I didn't add any unit tests right?

Comment (REVERB283) [reply packages/mermaid/src/diagrams/common/common.ts:35]: I'm getting Heap out of memory again and again when I run pnpm test, just how much memory do I need to allocate for the same.

Screenshot (28)

**Comment by REVERB283** [comment]
I don't understand the tasks that are failing and how to resolve those?

**Review by nirname — APPROVED** [review]

**Comment by nirname** [comment]
I have not investigated it further, but GA says
```
TypeError: Le.addHook is not a function
    at file:///home/runner/work/mermaid/mermaid/packages/mermaid/src/vitepress/.vitepress/.temp/app.js:8476:4
    at ModuleJob.run (node:internal/modules/esm/module_job:194:25)
```
perhaps something else is needed to make it happen, we cannot merge it while the error it the build is being reproduced

**Comment by REVERB283** [comment]
It could be that dompurify apis are not accessible when we're using the sanitzation on server end maybe, it's my speculation.

**Comment by nirname** [comment]
Yes, is it a maybe. Same guess. In that case we have to find a workaround

**Comment by nirname** [comment]
I am not well aware of DOMPurify functions, but `DOMPurify.sanitize` works

**Comment by REVERB283** [comment]
@nirname So I was digging in on this and I found out that people are facing issues while using dompurify with SSR and the alternative is to use either jsDom or isomorphic-dompurify you may refer to the following thread

[stackoverflow](https://stackoverflow.com/questions/65646007/next-js-dompurify-sanitize-shows-typeerror-dompurify-webpack-imported-module)

but it still gave me addhook is not defined

**Comment by nirname** [comment]
@REVERB283 Yes, I saw this issue too. And I am at the same point as you are now. Try investigating it further, please.

Let's also summon @sidharthv96, perhaps he will give us the right direction. To summarize: 

GA actions say `addHook` is not a function. We rely on the assumption that this should be available just as `sanitize`. But it may happen that `addHook` won't be available in SSR (if there is server side rendering). I have tried building docs locally and could not reproduce this error.

**Comment by REVERB283** [comment]
You can try using 
```
cd packages/mermaid && npm run docs:build:vitepress       
```

**Comment by REVERB283** [comment]
@nirname I managed to resolve the remaining issues, please review it once again.

**Review by nirname — APPROVED** [review]

**Comment by nirname** [comment]
@REVERB283 yes, I see. Changes are OK. I do not fully understand why this works. Could you explain it for me? I'd like to clarify it for myself.

**Comment by nirname** [comment]
@REVERB283 could you add some tests to ensure that no future requests will brake this feature?

**Comment by REVERB283** [comment]
> @REVERB283 yes, I see. Changes are OK. I do not fully understand why this works. Could you explain it for me? I'd like to clarify it for myself.

I too don't have the proper clarity myself but it seems like when I call addHook() in the global scope it doesn't seem to be accessible; could be possible that it gets called before DOMPurify gets imported/defined (which should be highly unlikely). However when I call addHook() inside removeScript() it seems to work without errors.

I've come to this conclusion after updating the code and calling the guthub workflow "build-docs" multiple times on my local system using "act".

**Comment by REVERB283** [comment]
> @REVERB283 could you add some tests to ensure that no future requests will brake this feature?

I'm very new to unit testing but I can surely try. Can you help me out with the files where I need to write the test cases? Also can you point out some already implemented examples that I can take reference from?

Much appreciated.

**Comment by nirname** [comment]
@REVERB283 yes, sure. There are 2 types of tests - unit tests, which we mainly use for testing parsers, because they are internal. And e2e (end-to-end) also known as integration tests. What they do is starting client and performing actions on behalf of it (like clicking links in headless chrome, sending requests and fetching responses). In our case they check what the rendered image is, so these are used to test renderers.

All the integration tests are inside [cypress folder](https://github.com/mermaid-js/mermaid/tree/develop/cypress/integration/rendering), like [this pie rendering test](https://github.com/mermaid-js/mermaid/blob/develop/cypress/integration/rendering/pie.spec.ts).

Tests for parsers are located in a specific diagram folder inside mermaid package, like this [pie parser test](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/diagrams/pie/pie.spec.ts).

Our current current e2e testing tool is Cypress. To reduce amount of checks and ensure that graph layout is stable we often compare 2 images. In our specific case we need to ensure that _target attribute inside SVG is present or absent. Have a look at [this example](https://github.com/mermaid-js/mermaid/blob/develop/cypress/integration/rendering/sankey.spec.ts#L27), which in fact should have been implemented with image comparison (my bad). Anyway, this is just an example, good or bad, which you can take as a base. Very likely image comparison won't work for this current fix.

[Here is another one](https://github.com/mermaid-js/mermaid/blob/develop/cypress/integration/rendering/flowchart-v2.spec.js#L100) for flowchart, which is basically identical. May be a little bit more descriptive since `imgSnapshotTest` and `cy.get('svg').should((svg) => {})` expectations are used there along each other.

**Comment by REVERB283** [comment]
@nirname I've added few test cases, can you review?

**Review by nirname — APPROVED** [review]
That'll do! May be we should wait for one more approval, but basically it is ready. Great job.

**Comment by REVERB283** [comment]
@sidharthv96 any update?

**Comment by huynhicode** [comment]
@REVERB283 - Just chiming in here to let you know that @sidharthv96 is currently on vacation.

**Pull request merged**
